### PR TITLE
fortran-fpm: init at 0.9.0

### PR DESCRIPTION
**Pull request by proofconstruction**
###### Description of changes
`fpm` is the package manager and build system for Fortran.

https://fpm.fortran-lang.org

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/package-management/fortran-fpm/default.nix
Comment (Janik-Haag) [inline pkgs/tools/package-management/fortran-fpm/default.nix:1]:
```suggestion
{ lib
, fetchurl
, pkgs
, stdenv 
}:
```

just a minor formatting nitpick

File: pkgs/tools/package-management/fortran-fpm/default.nix
Comment (Janik-Haag) [inline pkgs/tools/package-management/fortran-fpm/default.nix:29]:
```suggestion
  buildPhase = ''
    runHook preBuild
    
    mkdir -p ${buildPath}
    gfortran -J ${buildPath} -o ${buildPath}/${pname} $src
    
    runHook postBuild
  '';
```
formatting nitpick

File: pkgs/tools/package-management/fortran-fpm/default.nix
Comment (Janik-Haag) [inline pkgs/tools/package-management/fortran-fpm/default.nix:22]:
```suggestion
  nativeBuildInputs = with pkgs; [ 
    gfortran
  ];
```
formatting nitpick

File: pkgs/tools/package-management/fortran-fpm/default.nix
Comment (Janik-Haag) [inline pkgs/tools/package-management/fortran-fpm/default.nix:36]:
```suggestion
  installPhase = ''
    runHook preInstall
    
    mkdir -p $out/bin
    cp ${buildPath}/${pname} $out/bin
    
    runHook postInstall
  '';
```
formatting nitpick

**Review by Janik-Haag — COMMENTED** [review]
Mostly nitpicks, quite a good pr for your first time :heart:

**Comment by Janik-Haag** [comment]
Result of `nixpkgs-review pr 239035` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>fortran-fpm</li>
  </ul>
</details>

File: pkgs/tools/package-management/fortran-fpm/default.nix
Comment (proofconstruction) [reply pkgs/tools/package-management/fortran-fpm/default.nix]: Thanks for the info; do you have a link to discussion about this? I'd like to make sure we update the relevant parts of the nixpkgs manual with the style guide

File: pkgs/tools/package-management/fortran-fpm/default.nix
Comment (proofconstruction) [reply pkgs/tools/package-management/fortran-fpm/default.nix]: I moved pkgs out to the `with` clause but left this as a single line

File: pkgs/tools/package-management/fortran-fpm/default.nix
Comment (piegamesde) [reply pkgs/tools/package-management/fortran-fpm/default.nix]: IIRC, the motivation for using let bindings was because recursive attribute sets in `mkDerviation` had weird semantics with overrides. Not sure how let bindings made this better though.

Anyways, nowadays we can call `mkDerivation` with a recursive function instead, which does have the proper behavior with overrides.

**Review by Janik-Haag — APPROVED** [review]
LGTM; I will give it a few more days and if no one complains merge it.

**Comment by Janik-Haag** [comment]
Ah i missed your third commit :see_no_evil: can you squash the `fix rec` commit into the init commit? I will merge this once you have done that.

**Comment by proofconstruction** [comment]
@Janik-Haag Whoops, I figured everything would be squash-merged so I left it alone, but it's all set now.

**Comment by Janik-Haag** [comment]
No problem, I just checked the when approving and reviewed the entire PR before merging and that's when I noticed it ^^
Just waiting for ci now :D

**Pull request merged**
